### PR TITLE
Processing history rework

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR partially implements the framework described in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/130, and ports the processing history log over to this framework.

While there's no user visible changes which arise from this, it's a much more flexible framework as it allows us to store a full variant map in the history. So we can use this in future PRs to store more than just the python command in the history (e.g. storing the qgis_process command, the environment of the algorithm execution (ellipsoid settings, distance/area units, etc), AND any outputs given by the algorithm execution and the full output log.

History log storage in now stored in a sqlite database instead of the text file previously used, which will allow for more efficient storage/retrieval/filtering.

(This framework extends to use for more than just Processing -- it could also be used to store the history of other user operations, such as project edits, layout exports, ...)

Also now includes new right-click actions in the Processing history dialog to copy entries as JSON or qgis_process commands

Sponsored by the Research Institute for Nature and Forest, Flemish Govt

### CODE REVIEW

**Comment by florisvdh** [comment]
Not yet rebased on #46497 (so `qgis_process` doesn't yet accept JSON here), but copying the JSON string from history does work :rocket:!

Note: here it is a oneline JSON string, while it was a multiline JSON string in #46497. You may want to bring it into alignment - but anyway this works!

<details><summary><strong>Screenshot</strong></summary>

afbeelding

Using:

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /></head><body>


QGIS version | 3.23.0-Master | QGIS code revision | a2443ea4
-- | -- | -- | --
Qt version | 5.12.8
Python version | 3.8.10
GDAL/OGR version | 3.3.2
PROJ version | 7.2.1
EPSG Registry database version | v10.008 (2020-12-16)
GEOS version | 3.9.1-CAPI-1.14.2
SQLite version | 3.31.1
PostgreSQL client version | 12.9 (Ubuntu 12.9-0ubuntu0.20.04.1)
SpatiaLite version | 5.0.0
QWT version | 6.1.4
QScintilla2 version | 2.11.2
OS version | Linux Mint 20
  |   |   |  
Active Python plugins
geopunt4Qgis | 2.2.4
quick_map_services | 0.19.19
grassprovider | 2.12.99
processing | 2.12.99
sagaprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.5

</body></html>

</details>

**Comment by nyalldawson** [comment]
>  You may want to bring it into alignment - but anyway this works!

Good catch! I'll fix

File: python/plugins/processing/gui/HistoryDialog.py [hunk 1]
```diff
@@ -179,45 +187,114 @@ def fillTree(self):
...
 
     def createTest(self):
         item = self.tree.currentItem()
         if isinstance(item, TreeLogEntryItem):
-            if item.isAlg:
-                TestTools.createTest(item.entry.text)
+            if item.as_python_command():
+                TestTools.createTest(item.as_python_command())
+
+    def copy_text(self, text: str):
```
Comment (nirvn) [inline python/plugins/processing/gui/HistoryDialog.py:215]: For consistency's sake, should this be renamed to copyText?

File: src/gui/history/qgshistoryprovider.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/***************************************************************************
+                            qgshistoryprovider.cpp
+                            -------------------------
+    begin                : April 2019
```
Comment (nirvn) [inline src/gui/history/qgshistoryprovider.cpp:4]: Love a 2019 idea coming to fruition :)

**Review by nirvn — APPROVED** [review]
Nice foundation.

File: python/plugins/processing/gui/HistoryDialog.py (hunk 1 above)
Comment (nyalldawson) [reply python/plugins/processing/gui/HistoryDialog.py:215]: I just can't bring myself to use that capitalization style for Python code :wink:

File: src/gui/history/qgshistoryprovider.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/gui/history/qgshistoryprovider.cpp:4]: Heh... the actual QEP predates even that file!!

**Pull request merged**